driver/src/test/java/org/neo4j/driver/internal/handlers/HelloResponseHandlerTest.java

```diff
@@ -0,0 +1,135 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.neo4j.driver.internal.async.ChannelAttributes.serverVersion;
+import static org.neo4j.driver.internal.async.ChannelAttributes.setMessageDispatcher;
+import static org.neo4j.driver.internal.async.outbound.OutboundMessageHandler.NAME;
+import static org.neo4j.driver.internal.logging.DevNullLogging.DEV_NULL_LOGGING;
+import static org.neo4j.driver.v1.Values.value;
+
+class HelloResponseHandlerTest
+{
+    private final EmbeddedChannel channel = new EmbeddedChannel();
+
+    @BeforeEach
+    void setUp()
+    {
+        setMessageDispatcher( channel, new InboundMessageDispatcher( channel, DEV_NULL_LOGGING ) );
+        ChannelPipeline pipeline = channel.pipeline();
+        pipeline.addLast( NAME, new OutboundMessageHandler( new MessageFormatV1(), DEV_NULL_LOGGING ) );
+        pipeline.addLast( new ChannelErrorHandler( DEV_NULL_LOGGING ) );
+    }
+
+    @AfterEach
+    void tearDown()
+    {
+        channel.finishAndReleaseAll();
+    }
+
+    @Test
+    void shouldSetServerVersionOnChannel()
+    {
+        ChannelPromise channelPromise = channel.newPromise();
+        HelloResponseHandler handler = new HelloResponseHandler( channelPromise );
+
+        Map<String,Value> metadata = singletonMap( "server", value( ServerVersion.v3_2_0.toString() ) );
+        handler.onSuccess( metadata );
+
+        assertTrue( channelPromise.isSuccess() );
+        assertEquals( ServerVersion.v3_2_0, serverVersion( channel ) );
+    }
+
+    @Test
+    void shouldThrowWhenServerVersionNotReturned()
+    {
+        ChannelPromise channelPromise = channel.newPromise();
+        HelloResponseHandler handler = new HelloResponseHandler( channelPromise );
+
+        assertThrows( IllegalStateException.class, () -> handler.onSuccess( singletonMap( "server", null ) ) );
+
+        assertFalse( channelPromise.isSuccess() ); // initialization failed
+        assertTrue( channel.closeFuture().isDone() ); // channel was closed
+    }
+
+    @Test
+    void shouldThrowWhenServerVersionInNull()
+    {
+        ChannelPromise channelPromise = channel.newPromise();
+        HelloResponseHandler handler = new HelloResponseHandler( channelPromise );
+
+        assertThrows( IllegalStateException.class, () -> handler.onSuccess( singletonMap( "server", Values.NULL ) ) );
+
+        assertFalse( channelPromise.isSuccess() ); // initialization failed
+        assertTrue( channel.closeFuture().isDone() ); // channel was closed
+    }
+
+    @Test
+    void shouldThrowWhenServerVersionCantBeParsed()
+    {
+        ChannelPromise channelPromise = channel.newPromise();
+        HelloResponseHandler handler = new HelloResponseHandler( channelPromise );
+
+        assertThrows( IllegalArgumentException.class, () -> handler.onSuccess( singletonMap( "server", value( "WrongServerVersion" ) ) ) );
```

Comment:
Is it still required to have the server version parsed rather than simply record it as a simple message?

Reply:
Server version should not be needed for version comparisons in Bolt V3. It is just easier to still keep it as `ServerVersion` object for older protocol versions